### PR TITLE
Expec p th

### PR DESCRIPTION
**Pull request by wcbeard**
Adds functionality to compute expected value of latent variabls p and theta.

### CODE REVIEW

File: bgbb/bgbb_utils.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    n_opps_name="n_opps",
+    seed=0,
+):
+    """
+    """
+    nr.seed(seed)
+    if p_th is None:
+        raise NotImplementedError
+        # to dict
+        p, th = _gen_probs(abgd, n=n_users)
```
Comment (acmiyaguchi) [inline bgbb/bgbb_utils.py]: This code pathway looks unreachable.

File: bgbb/bgbb_utils.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    for opp, buy, die in zip(range(1, len(buys) + 1), buys, dies):
+        if die:
+            break
+        if buy:
+            x += 1
+            tx = opp
+    return x, tx
+
+
+class AbgdParams:
```
Comment (acmiyaguchi) [inline bgbb/bgbb_utils.py]: I think you might be looking for the namedtuple collection here. It lets you define a tuple with names.

```python
from collections import namedtuple

AbgdParams = namedtuple("AbgdParams", ["a", "b", "g", "d"])

# constructing a new tuple
x = AbgdParams(1, 2, 3, 4)

# desugar
a, b, g, d = x

print(a, b, g, d)
# (1, 2, 3, 4)

# return an ordered dictionary
x._asdict()
# OrderedDict([('a', 1), ('b', 2), ('g', 3), ('d', 4)])

File: bgbb/core.py [hunk 3]
```diff
@@ -113,3 +119,27 @@ def cond_exp_rets_till_nb(self, t, frequency, recency, n, params):
...
+        l_g_frac = np.log(g) - np.log(g + d)
+
+        return (l_a_frac + l_ap1 - ll), (l_g_frac + l_gp1 - ll)
+
+    @classmethod
+    def _expec_p_th(cls, a, b, g, d, f, r, n):
+        lp_exp, lth_exp = cls._log_expec_p_th(a, b, g, d, f, r, n)
+        return np.exp(lp_exp), np.exp(lth_exp)
+
+    def expec_p_th(self, f, r, n) -> Tuple[Sequence[Prob], Sequence[Prob]]:
```
Comment (acmiyaguchi) [inline bgbb/core.py]: As a public method, it would be nice to have a readable name like `get_latent_variables`.

File: bgbb/bgbb_utils.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
+from collections import OrderedDict
+from itertools import count
+from typing import List, Optional, Sequence, Tuple
+
+import numpy as np
+
+# sim_dat = gen_buy_die(n_opps, n_users, abgd=SimPars.lst, p_th=[pp, th])
```
Comment (acmiyaguchi) [inline bgbb/bgbb_utils.py]: nit: remove comment

File: bgbb/bgbb_utils.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+def gen_buy_die(
+    n_opps,
+    n_users,
+    abgd: List[float],
+    p_th: Optional[Tuple[Sequence[Prob], Sequence[Prob]]] = None,
+    n_opps_name="n_opps",
+    seed=0,
+):
+    """
+    """
```
Comment (acmiyaguchi) [inline bgbb/bgbb_utils.py:36]: This is a good opportunity to fill out the documentation for this function.

**Review by acmiyaguchi — APPROVED** [review]
This looks reasonable to me -- the simulation for testing latent variable recovery is nice.

There are a few nits and comments spread about. In particular, you may want to play with namedtuple to see if it fits your use-case.

File: bgbb/bgbb_utils.py (hunk 1 above)
Comment (wcbeard) [reply bgbb/bgbb_utils.py]: Got rid of NotImplementedError and made test for this case

File: bgbb/core.py (hunk 3 above)
Comment (wcbeard) [reply bgbb/core.py]: changed to `latent_variable_mean`

**Pull request merged**